File: js/src/core/e2ee-request.ts
import * as v from 'valibot';
import { ChatCompletionRequestSchema } from '../schemas';
import type {
  PreparedE2eeChatRequest,
  PrepareE2eeChatRequestParams,
} from '../types/e2ee';
import type { SecureChatCompletionResponse } from '../types/inference-client';
import { ApiError } from '../utils/errors';
import { NO_ALIASING_HEADER } from './cloud-api';
import { createE2eeClientKeyPair, type E2eeClientKeyPair } from './e2ee';
import {
  createE2eeChatSseTransform,
  decryptE2eeChatResponse,
  encryptE2eeChatRequest,
  parseE2eeChatResponse,
} from './e2ee-chat';

type DecodeChatRequestParams = {
  readonly request: Request;
};

type DecryptE2eeResponseParams = {
  readonly response: Response;
  readonly clientKeyPair: E2eeClientKeyPair;
};

type AwaitWithAbortParams<T> = {
  readonly operation: Promise<T>;
  readonly signal: AbortSignal;
};

/**
 * Prepare one Chat Completions request using the supplied model E2EE key.
 *
 * Each call creates a fresh response key pair. The returned decryption
 * operation handles JSON and SSE without exposing the private key. Only
 * protocol-defined Chat fields are encrypted; other JSON is preserved.
 * The caller verifies the key's attestation before preparing the request.
 * This operation performs no network requests or completion-signature checks.
 */
export async function prepareE2eeChatRequest({
  request,
  modelKey,
}: PrepareE2eeChatRequestParams): Promise<PreparedE2eeChatRequest> {
  request.signal.throwIfAborted();
  if (request.method !== 'POST') {
    throw invalidInput({
      field: 'request',
      reason: 'unsupported_value',
      expected: 'a POST Chat Completions request',
      actual: request.method,
    });
  }
  const value = await decodeChatRequest({ request });
  const parsed = v.safeParse(ChatCompletionRequestSchema, value);
  if (!parsed.success) {
    throw invalidInput({
      field: 'request body',
      reason: 'unsupported_value',
      expected: 'a JSON Chat Completions request with a string model',
    });
  }
  request.signal.throwIfAborted();
  const clientKeyPair = createE2eeClientKeyPair(modelKey.signingAlgo);
  const encrypted = encryptE2eeChatRequest({ body: parsed.output, modelKey });
  const headers = new Headers(request.headers);
  headers.delete('content-length');
  headers.set('content-type', 'application/json');
  removeE2eeHeaders(headers);
  headers.set('x-signing-algo', modelKey.signingAlgo);
  headers.set('x-client-pub-key', clientKeyPair.publicKey);
  headers.set('x-model-pub-key', modelKey.publicKey);
  if (modelKey.signingAlgo === 'ed25519') {
    headers.set('x-encryption-version', '2');
  }
  headers.set(NO_ALIASING_HEADER, 'true');
  headers.set('x-encrypt-all-fields', 'true');

  return {
    request: new Request(request, {
      headers,
      body: JSON.stringify(encrypted.body),
    }),
    decryptResponse: (response) => decryptResponse({ response, clientKeyPair }),
  };
}

/** Internal Chat body reader shared by the public preparation and client paths. */
export async function decodeChatRequest({
  request,
}: DecodeChatRequestParams): Promise<unknown> {
  request.signal.throwIfAborted();
  let reader: ReadableStreamDefaultReader<Uint8Array> | undefined;
  const decoder = new TextDecoder();
  let text = '';
  try {
    reader = request.clone().body?.getReader();
    if (reader !== undefined) {
      while (true) {
        const chunk = await awaitWithAbort({
          operation: reader.read(),
          signal: request.signal,
        });
        if (chunk.done) break;
        text += decoder.decode(chunk.value, { stream: true });
      }
      text += decoder.decode();
    }
    return JSON.parse(text);
  } catch (cause) {
    if (request.signal.aborted) {
      // A cloned stream's cancellation may wait for the other branch. Preserve
      // the caller's abort immediately instead of awaiting that cancellation.
      void reader?.cancel(request.signal.reason).catch(() => undefined);
      throw request.signal.reason;
    }
    throw invalidInput(
      {
        field: 'request body',
        reason: 'invalid_json',
        expected: 'a JSON Chat Completions request',
      },
      cause,
    );
  } finally {
    reader?.releaseLock();
  }
}

export function isServerSentEventResponse(response: Response): boolean {
  return isServerSentEventContentType(response.headers.get('content-type'));
}

export function isServerSentEventContentType(
  contentType: string | null,
): boolean {
  return contentType?.toLowerCase().startsWith('text/event-stream') ?? false;
}

export function removeE2eeHeaders(headers: Headers): void {
  headers.delete('x-signing-algo');
  headers.delete('x-client-pub-key');
  headers.delete('x-model-pub-key');
  headers.delete('x-encryption-version');
  headers.delete('x-encrypt-all-fields');
}

async function decryptResponse({
  response,
  clientKeyPair,
}: DecryptE2eeResponseParams): Promise<Response> {
  if (!response.ok) return response;
  const headers = new Headers(response.headers);
  headers.delete('content-length');
  if (isServerSentEventResponse(response)) {
    if (response.body === null) throw invalidResponse();
    return new Response(
      response.body.pipeThrough(createE2eeChatSseTransform({ clientKeyPair })),
      { status: response.status, statusText: response.statusText, headers },
    );
  }
  const body = await decodeResponseBody(response);
  const decrypted = decryptE2eeChatResponse({ body, clientKeyPair });
  headers.set('content-type', 'application/json');
  return new Response(JSON.stringify(decrypted), {
    status: response.status,
    statusText: response.statusText,
    headers,
  });
}

async function decodeResponseBody(
  response: Response,
): Promise<SecureChatCompletionResponse> {
  let text: string;
  try {
    text = await response.text();
  } catch (cause) {
    throw new ApiError(
      {
        code: 'api.transport_failed',
        details: { resource: 'completion', reason: 'response_body' },
        retryable: true,
      },
      { cause },
    );
  }
  let value: unknown;
  try {
    value = JSON.parse(text);
  } catch (cause) {
    throw invalidResponse(cause);
  }
  return parseE2eeChatResponse({ body: value });
}

function awaitWithAbort<T>({
  operation,
  signal,
}: AwaitWithAbortParams<T>): Promise<T> {
  if (signal.aborted) return Promise.reject(signal.reason);
  return new Promise((resolve, reject) => {
    const abort = (): void => reject(signal.reason);
    signal.addEventListener('abort', abort, { once: true });
    void operation.then(
      (value) => {
        signal.removeEventListener('abort', abort);
        resolve(value);
      },
      (cause: unknown) => {
        signal.removeEventListener('abort', abort);
        reject(cause);
      },
    );
  });
}

function invalidInput(
  details: Extract<
    ApiError['failure'],
    { code: 'api.invalid_input' }
  >['details'],
  cause?: unknown,
): ApiError {
  return new ApiError(
    { code: 'api.invalid_input', details },
    cause === undefined ? undefined : { cause },
  );
}

function invalidResponse(cause?: unknown): ApiError {
  return new ApiError(
    {
      code: 'api.invalid_response',
      details: {
        path: 'Chat Completions response',
        expected: 'an encrypted Chat Completions response',
        actual: 'invalid',
      },
    },
    cause === undefined ? undefined : { cause },
  );
}

File: py/src/nearai_inference_sdk/core/e2ee_request.py
"""Standalone encrypted Chat requests with transport-independent decryption."""

from __future__ import annotations

import json
from collections.abc import AsyncIterator
from typing import Any

import httpx
from pydantic import ValidationError

from ..schemas import ChatCompletionRequestSchema
from ..types.e2ee import E2eeModelKey, PreparedE2eeChatRequest
from ..utils.errors import api_failure
from .e2ee import E2eeClientKeyPair, create_e2ee_client_key_pair
from .e2ee_chat import (
    decrypt_e2ee_chat_response,
    decrypt_e2ee_chat_sse,
    encrypt_e2ee_chat_request,
    parse_e2ee_chat_response,
)


_REPLACED_BODY_HEADERS = (
    'content-length',
    'transfer-encoding',
    'trailer',
    'content-md5',
    'digest',
    'content-digest',
    'repr-digest',
    'content-encoding',
)


async def prepare_e2ee_chat_request(
    request: httpx.Request, model_key: E2eeModelKey
) -> PreparedE2eeChatRequest:
    """Encrypt Chat fields for an already-verified key; perform no network I/O.

    Every call creates a fresh response key. The returned decryptor handles JSON
    and streaming SSE without exposing that private key. Keep the original wire
    request and response bytes separately when verifying a completion signature.
    """

    if request.method != 'POST':
        raise api_failure(
            'api.invalid_input',
            {
                'field': 'request',
                'reason': 'unsupported_value',
                'expected': 'a POST Chat Completions request',
                'actual': request.method,
            },
        )
    body = await decode_chat_request(request)
    client_key_pair = create_e2ee_client_key_pair(model_key.signing_algo)
    encrypted = encrypt_e2ee_chat_request(body, model_key)
    headers = httpx.Headers(request.headers)
    for name in _REPLACED_BODY_HEADERS:
        headers.pop(name, None)
    headers['content-type'] = 'application/json'
    remove_e2ee_headers(headers)
    headers['x-signing-algo'] = model_key.signing_algo
    headers['x-client-pub-key'] = client_key_pair.public_key
    headers['x-model-pub-key'] = model_key.public_key
    if model_key.signing_algo == 'ed25519':
        headers['x-encryption-version'] = '2'
    headers['x-no-aliasing'] = 'true'
    headers['x-encrypt-all-fields'] = 'true'

    async def decrypt_response(response: httpx.Response) -> httpx.Response:
        return await _decrypt_response(response, client_key_pair)

    return PreparedE2eeChatRequest(
        request=httpx.Request(
            request.method,
            request.url,
            headers=headers,
            json=encrypted,
            extensions=dict(request.extensions),
        ),
        decrypt_response=decrypt_response,
    )


async def decode_chat_request(request: httpx.Request) -> dict[str, Any]:
    """Read and validate the external HTTP request's minimal Chat boundary."""

    try:
        body = json.loads((await request.aread()).decode('utf-8'))
    except (ValueError, UnicodeError) as cause:
        raise api_failure(
            'api.invalid_input',
            {'field': 'request body', 'reason': 'invalid_json'},
            cause=cause,
        ) from cause
    try:
        return ChatCompletionRequestSchema.model_validate(body).model_dump()
    except ValidationError as cause:
        raise api_failure(
            'api.invalid_input',
            {
                'field': 'request body',
                'reason': 'unsupported_value',
                'expected': 'a JSON Chat Completions request with a string model',
            },
            cause=cause,
        ) from cause


def is_server_sent_event_response(response: httpx.Response) -> bool:
    return (
        response.headers.get('content-type', '').lower().startswith('text/event-stream')
    )


def remove_e2ee_headers(headers: httpx.Headers) -> None:
    for name in (
        'x-signing-algo',
        'x-client-pub-key',
        'x-model-pub-key',
        'x-encryption-version',
        'x-encrypt-all-fields',
    ):
        headers.pop(name, None)


async def _decrypt_response(
    response: httpx.Response, client_key_pair: E2eeClientKeyPair
) -> httpx.Response:
    if not response.is_success:
        return response
    headers = httpx.Headers(response.headers)
    # Decryption invalidates the wire length and digests. HTTPX already
    # decompresses source bytes, so the transformed body must not be decoded again.
    for name in _REPLACED_BODY_HEADERS:
        headers.pop(name, None)
    try:
        request = response.request
    except RuntimeError:
        request = None
    options = dict(
        status_code=response.status_code,
        headers=headers,
        extensions=dict(response.extensions),
        request=request,
    )
    if is_server_sent_event_response(response):
        return httpx.Response(
            **options, stream=_DecryptedSseStream(response, client_key_pair)
        )
    try:
        raw = await response.aread()
    except httpx.HTTPError as cause:
        raise api_failure(
            'api.transport_failed',
            {'resource': 'completion', 'reason': 'response_body'},
            retryable=True,
            cause=cause,
        ) from cause
    finally:
        await response.aclose()
    try:
        body = json.loads(raw.decode('utf-8'))
    except (ValueError, UnicodeError) as cause:
        raise api_failure(
            'api.invalid_response',
            {
                'path': 'Chat Completions response',
                'expected': 'a JSON Chat Completions object',
                'actual': 'invalid JSON',
            },
            cause=cause,
        ) from cause
    parsed = parse_e2ee_chat_response(body)
    decrypted = decrypt_e2ee_chat_response(parsed, client_key_pair)
    headers['content-type'] = 'application/json'
    return httpx.Response(**options, json=decrypted)


class _DecryptedSseStream(httpx.AsyncByteStream):
    def __init__(self, response: httpx.Response, client_key_pair: E2eeClientKeyPair):
        self.response = response
        self.client_key_pair = client_key_pair

    async def __aiter__(self) -> AsyncIterator[bytes]:
        try:
            async for chunk in decrypt_e2ee_chat_sse(
                self.response.aiter_bytes(), self.client_key_pair
            ):
                yield chunk
        finally:
            await self.response.aclose()

    async def aclose(self) -> None:
        await self.response.aclose()
